### PR TITLE
#1360 Implement POST /v3/notifications, including tests and API specifications

### PR DESCRIPTION
**Pull request by kalbfled**
# Description

Implement POST /v3/notifications/<email|sms> , including tests and API specifications.  These changes only implement the schema validation and route handling.  Actually sending the notifications will be implemented under #1361.

v3 is behind a feature flag that these changes enable for local, Github, Dev, and Perf.  Requests to these routes should generate a 404 in staging and production.

One change from v2 is that the v3 endpoints should accept contact information (phone number or e-mail address) and a recipient ID.  v2 is one or the other; not both.  I have attempted to comment my code sufficiently to explain my design decisions, so I am not listing them all here.

These changes include OpenAPI and Postman additions for v3.  You can import both into Postman and the former into [Swagger Editor](https://editor-next.swagger.io/).

These changes also add a new "ignore" to Makefile for the "certifi" Python package that should be removed during the next routine dependencies upgrade.

#1360 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Be sure to add "V3_ENABLED=True" to your local .docker-env file.  See the change in .docker-env.example.

- [x] All unit tests pass
- [x] [Deployed](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/6016838682) to Dev
- [x] Tested with Postman

## Checklist

- [x] Appropriate labels added to this pull request (i.e. "enhancement", "dependencies", etc.)
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -385,6 +388,26 @@ def register_v2_blueprints(application):
...
+    """
+
+    # TODO 1360 - require auth
+    # from app.authentication.auth import validate_service_api_key_auth
+    from app.v3 import v3_blueprint
+
+    # Notifications
+    from app.v3.notifications.rest import v3_notifications_blueprint
+    # TODO 1360 - require auth
+    # v3_notifications_blueprint.before_request(validate_service_api_key_auth)
```
Comment (kalbfled) [inline app/__init__.py:405]: Uncommenting these lines causes the /v3/notification/* routes to return a 403 error.

File: app/models.py
Comment (kalbfled) [inline app/models.py:1]: EMAIL_TYPE was declared twice.

File: app/v3/notifications/rest.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    request_data = request.get_json()
+    request_data["notification_type"] = EMAIL_TYPE if request.base_url.endswith("email") else SMS_TYPE
+
+    # This might raise jsonschema.ValidationError, which should trigger an error handler in app/v3/__init__.py.
+    v3_notifications_post_request_validator.validate(request_data)
+
+    request_data["id"] = uuid4()
+    # TODO 1361 - Pass the validated request data to Celery by calling apply_async (or something else)
+
+    return {"id": request_data["id"]}, 202
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply app/v3/notifications/rest.py:29]: "id" has been validated as a UUID, which cannot be malicious.

File: app/v3/notifications/rest.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from app.v3.notifications.notification_schemas import notification_v3_post_request_schema
+from flask import Blueprint, request
+from jsonschema.validators import Draft202012Validator
+from uuid import uuid4
+
+v3_notifications_blueprint = Blueprint("v3_notifications", __name__, url_prefix='/notifications')
+
+# Create an instance of Draft202012Validator to validate post request data.  Unit tests
+# should call jsonschema.validate, which also verifies that the schema is valid.
+v3_notifications_post_request_validator = Draft202012Validator(notification_v3_post_request_schema)
```
Comment (kalbfled) [inline app/v3/notifications/rest.py:13]: This is outside the route handler so it doesn't get recreated with every request.  The goal is to minimize response time.

File: app/models.py [hunk 4]
```diff
@@ -37,10 +37,11 @@
 from sqlalchemy.orm.collections import attribute_mapped_collection, InstrumentedList
 
 
-SMS_TYPE = 'sms'
 EMAIL_TYPE = 'email'
 LETTER_TYPE = 'letter'
```
Comment (cris-oddball) [inline app/models.py:41]: do we really need LETTER? it's going to be pulled out of v2 shortly.

File: app/models.py [hunk 5]
```diff
@@ -607,9 +608,6 @@ def __repr__(self):
         return '<{} has service permission: {}>'.format(self.service_id, self.permission)
```
Comment (cris-oddball) [inline app/models.py:608]: while we are here, can we just convert this to an f-string?

Comment (kalbfled) [reply app/models.py:608]: Happily, but I am the guy people like to cite as somebody who introduces unrelated changes . . .

File: app/v3/notifications/rest.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+
+@v3_notifications_blueprint.route("/email", methods=["POST"], endpoint="v3_notification_email")
+@v3_notifications_blueprint.route("/sms", methods=["POST"], endpoint="v3_notification_sms")
+def post_notification_v3():
+    request_data = request.get_json()
+    request_data["notification_type"] = EMAIL_TYPE if request.base_url.endswith("email") else SMS_TYPE
+    post_notification_v3_validate_and_send_to_celery(request_data)
+    return {"id": request_data["id"]}, 202
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply app/v3/notifications/rest.py:22]: "id" has already been validated us a valid UUID.

File: app/v3/notifications/rest.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+
+@v3_notifications_blueprint.route("/email", methods=["POST"])
+def v3_post_notification_email():
+    request_data = request.get_json()
+    request_data["notification_type"] = EMAIL_TYPE
+
+    # This might raise jsonschema.ValidationError, which should trigger an error handler in
+    # app/v3/__init__.py that returns a 400 response.
+    return {"id": v3_send_notification(request_data)}, 202
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply app/v3/notifications/rest.py:23]: The data is validated against a schema.

File: app/v3/notifications/rest.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+
+@v3_notifications_blueprint.route("/sms", methods=["POST"])
+def v3_post_notification_sms():
+    request_data = request.get_json()
+    request_data["notification_type"] = SMS_TYPE
+
+    # This might raise jsonschema.ValidationError, which should trigger an error handler in
+    # app/v3/__init__.py that returns a 400 response.
+    return {"id": v3_send_notification(request_data)}, 202
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply app/v3/notifications/rest.py:33]: The data is validated against a schema.

File: documents/openapi/openapi.yaml [hunk 9]
```diff
@@ -1334,6 +1334,58 @@ paths:
           $ref: '#/components/responses/NotFound'
         '500':
           $ref: '#/components/responses/InternalServerError'
+  /v3/notifications/email:
```
Comment (cris-oddball) [inline documents/openapi/openapi.yaml:1337]: There are OpenAPI changes that need to be made here.

Right now, this schema only states that the id_type and id_value are the only required fields beside template_id. That is incorrect. 

**v3 as written**
Screen Shot 2023-08-30 at 7 49 29 AM

- note that anyOf has nothing in it
- note that email_address is not required
- note that recipient_identifier is not required
- note that id_type and id_value are required
- there are two errors in the openapi validator for lines 2339 and 2386
Screen Shot 2023-08-30 at 7 52 00 AM


**v2 with more accurate description**
Here, anyOf is used correctly - one could use email_address or recipient_identifier or both.
Note too that there are examples for using just email or just recipient_identifier in a dropdown - v3 should also have this and an example for including both
Screen Shot 2023-08-30 at 7 53 11 AM

Please update.

**Review by cris-oddball — CHANGES_REQUESTED** [review]

**Comment by cris-oddball** [comment]
Why does the new error message for 401 include an array in "message"? The current routes do not do this and I think it is easier for a client to parse a string rather than an array of strings.

Screen Shot 2023-08-30 at 8 22 10 AM

File: documents/openapi/openapi.yaml (hunk 9 above)
Comment (kalbfled) [reply documents/openapi/openapi.yaml:1337]: Updated.  Turns out json-schema and Swagger syntaxes are not exactly the same.

**Review by nikolai-efimov — DISMISSED** [review]
please add V3_ENABLED=True to ci/docker-compose-local-readwrite.yml as well.

**Comment by kalbfled** [comment]
> please add V3_ENABLED=True to ci/docker-compose-local-readwrite.yml as well.

@nikolai-efimov The file uses the variables in .docker-env, which is not version controlled.  You need to add it to your local file.  I did add V3_ENABLED to the template, .docker-env.example.

**Review by cris-oddball — APPROVED** [review]

**Review by nikolai-efimov — APPROVED** [review]

**Pull request merged**